Bundle our frontend via esbuild

mariusandra opened the pull request:
## Changes

- Replaces WebPack with [esbuild](https://esbuild.github.io/) for some 🔥 🔥 🔥 🚒 🚒 🚒 🦄 🌈 🚅 and "how is this even possible?".
- Deprecates https://github.com/PostHog/posthog/pull/6550

- Cold **full build in ~5 seconds** on a M1 (vs 150 seconds for webpack)
![2021-11-01 00 28 25](https://user-images.githubusercontent.com/53387/139604797-6efe66ee-6906-420f-abbb-1d7adfe47cea.gif)

- Cold boot to interactive in 10 seconds (vs 120 seconds for webpack)
- Full page reload in 3-4 seconds. There's no react "fast refresh" since that requires babel, which adds +30sec to everything. Webpack's full page reloads took a lot longer. 

![2021-11-01 00 14 04](https://user-images.githubusercontent.com/53387/139604469-4f92f310-bc18-4740-afc0-8db11cb509d1.gif)

## How did you test this code?

- Ran `esbuild` locally, thought it didn't do anything, but it actually built everything in about as time as it takes webpack to boot up. 
- This is game changing. Think of the speedups in CI alone.

## Caveats

- Use http://localhost:8234/ when testing to enable live reloads. It's also a drop-in replacement for the previous dev server, so http://localhost:8000/ works as well. Webpack's dev server on that port just returned JS, this one proxies other requests as well.
- I had to build a new dev server since esbuild doesn't come with any, and nothing available met our needs. They would have been much slower in reloads, and wouldn't support multiple build outputs. This dev server still needs tweaks, with the biggest bug being that it sometimes doesn't reload the static files on time. I think I need to replace the `live-server` package with a custom express app as well, but that felt like taking the project a bit too far.
- All requests on the dev server that are not to the built files, nor to django endpoints, are proxied to django's `index.html` to get the app context json (current user, etc) for a faster bootup.
- Since this replaces babel, we get rid of [babel-plugin-kea](https://github.com/keajs/babel-plugin-kea), which automatically injected paths onto logics. We'll have to add them manually for every logic (separate PR) or pick between a codemod/eslint/codegen/required-type/manual approach. We can also not add them, but then every logic will be `kea.logic.123` in the store, and debugging will be harder.
- Memory usage went down, but not by much. esbuild is now using 1.5GB for me (vs 2.11GB for webpack). It'd use ~600MB if I'd build the shared dashboards and toolbar only on start, and just watched the app for changes.
- This imports JS files in a slightly different order and style than webpack. Search for `keyMappingKeys` below, to see an example of something I had to move closer to the `export`, since in global scope after an `import` it was otherwise `undefined`. This is because e.g. the toolbar code is pushed to one `toolbar.js` file, and the modules are just loaded in some order, but that's undefined. Luckily there's always some workaround possible.
`index.html`
- AntD is currently fully built in. That means ~500kb of CSS, and a lot of modules. I tried adding a plugin to rewrite the imports to use `from 'antd/lib/bla'` instead of `from 'antd'`, but that added ~30sec to the build time (had babel?), so need to find a different solution.
- We're fully [using ES `import` statements](https://caniuse.com/?search=import) in the browser now. So some old things that worked before are no longer supported, but should be fine.

## Next steps

- ~~I need to see how well this performs in production. `esbuild`'s code splitting is still experimental, and does produce a lot of chunks. Perhaps with HTTP2 this is not that bad, but I just don't know.~~ It's working great, see below.

![2021-11-01 01 00 29](https://user-images.githubusercontent.com/53387/139605709-2b0fa127-8766-4f41-b29d-8103c70a5994.gif)

- Clean up and release this PR in chunks
- Make a PR to add paths to kea logics.

mariusandra committed:
esbuild package

mariusandra committed:
almost get esbuild working

mariusandra committed:
fix react-virtualized imports

mariusandra committed:
add splitting

mariusandra committed:
fix funny import reorder bug

mariusandra committed:
fix squeakAudio referring to itself

mariusandra committed:
write index.html file

mariusandra committed:
fix some bad imports

mariusandra committed:
update antd paths

mariusandra committed:
remove raw-loader usage, it didn't work anyway

mariusandra committed:
refactor and copy public

mariusandra committed:
build app and toolbar

mariusandra committed:
get toolbar working, but without styles

mariusandra committed:
make toolbar and its styles work

mariusandra committed:
shared dashboards

mariusandra committed:
clean frontend build before rebuilding

mariusandra committed:
add watch mode

mariusandra committed:
reorder tasks

mariusandra committed:
revert js url

mariusandra committed:
incremental builds of app with debounced chokidar watching

mariusandra committed:
common build/watch script

mariusandra committed:
improve logs

mariusandra committed:
watch during firrst build

mariusandra committed:
fix toolbar url

mariusandra committed:
fix wrongly exported scene

mariusandra committed:
create sceneProxyLogic to untangle sceneLogic from all bundles

mariusandra committed:
disconnect sceneLogic and refactor setPageTitle

mariusandra committed:
live reloading server

mariusandra committed:
rename utils file

mariusandra committed:
only wait for /static

mariusandra committed:
fix encoding

mariusandra committed:
simplify

mariusandra committed:
add missing dayjs plugins

mariusandra committed:
fix pathless logics

mariusandra committed:
simplify options

mariusandra committed:
add jsx for webapck

mariusandra committed:
slight delay to catch changes

mariusandra committed:
a type is a type

mariusandra committed:
fix build

mariusandra committed:
esbuild in start

mariusandra committed:
funnelLogic path

mariusandra commented:
Building the frontend in CI (task "docker test image build") takes 14 sec now:

![image](https://user-images.githubusercontent.com/53387/139671953-67258f43-851b-4be0-bd14-41be768966ca.png)

Before it was 10-11 min:

![image](https://user-images.githubusercontent.com/53387/139672144-0f0df835-096f-4548-8405-8d570d7533b0.png)

(why so slow? especially compared to cypress)

It'll also take 2-3 min off each cypress test:

![image](https://user-images.githubusercontent.com/53387/139678368-685b4c3d-519b-4517-8e4a-b519ddf51d52.png)

instead of:

![image](https://user-images.githubusercontent.com/53387/139678435-00664937-f6ad-4874-9d07-6410225b8829.png)



Combined with https://github.com/PostHog/posthog/pull/6759 this might make a lot of lives easier.

mariusandra committed:
include all files with a "." (so .mjs, etc) in /frontend/ to docker

mariusandra commented:
I built a docker image `mariusandra/posthog-esbuild:latest` with the last state of this PR, and deployed it on a personal self-hosted instance.

With "fast 3g" throttling, both apps load in about 10sec:

- webpack with "fast 3g" throttling:
![2021-11-01 16 09 35](https://user-images.githubusercontent.com/53387/139701357-f02cefed-fc6d-4ae8-9bf6-b0b6d9c5a535.gif)

- esbuild with "fast 3g" throttling:
![2021-11-01 16 53 32](https://user-images.githubusercontent.com/53387/139701238-01389425-e572-400a-ad32-d1d9d6262401.gif)

Without this throttling, both apps load virtually instantly for me, in about 2sec:
![2021-11-01 16 58 17](https://user-images.githubusercontent.com/53387/139701776-32eb82e9-444e-4516-83f4-1771a4618795.gif)

The nr of bytes loaded is even smaller with esbuild (~4MB vs ~5MB), and even if there are so many tiny ".js" chunks to load, all of them are requested immediately when loading a scene, not in a cascading order with multiple round trips. Hence HTTP2 batches them all together into not more roundtrips than we have now.

Hence: I say this is a go for production usage!

mariusandra committed:
rename to "utils.mjs", make "build.mjs" executable

mariusandra committed:
improve erroring

mariusandra committed:
revert some needless changes

mariusandra committed:
more reverts

mariusandra committed:
change some scripts

mariusandra committed:
remove setuff

mariusandra committed:
clarify function

mariusandra committed:
make "--host 0.0.0.0" work

mariusandra committed:
fix import order issue in webpack

mariusandra committed:
remove webpack css inlining for toolbar to simplify config

mariusandra committed:
make toolbar with external styles work in storybook

mariusandra commented on frontend/src/lib/utils/eventUsageLogic.ts:
```diff
@@ -24,11 +24,9 @@ import {
 import { Dayjs } from 'dayjs'
 import { preflightLogic } from 'scenes/PreflightCheck/logic'
-import { PersonModalParams } from 'scenes/trends/personsModalLogic'
+import type { PersonModalParams } from 'scenes/trends/personsModalLogic'
```
Just trying it out here, but we could add `type` everywhere we only import types, e.g. all logic types. This will tell esbuild to totally ignore these files when bundling, as types are ignored anyway. Kea-typegen should do this automatically.

mariusandra commented on frontend/src/lib/utils/eventUsageLogic.ts:
```diff
@@ -147,7 +145,7 @@ function sanitizeFilterParams(filters: Partial<FilterType>): Record<string, any>
     eventUsageLogicType<DashboardEventSource, GraphSeriesAddedSource, RecordingWatchedSource>
 >({
-    connect: [preflightLogic],
+    connect: () => [preflightLogic],
```
Had to add this to make these files now work with webpack. This is just `undefined` when the `kea({})` line in the global scope is executed.

mariusandra commented on frontend/src/toolbar/Toolbar.stories.tsx:
```diff
@@ -20,18 +20,20 @@ const toolbarStory = (json: Record<string, any>): (() => JSX.Element) => {
     return keaStory(() => {
         useEffect(() => {
-            const styleTags: HTMLStyleElement[] = Array.from(document.getElementsByTagName('style'))
-            ;(window as any)['__PHGTLB_STYLES__'] = styleTags.map((tag) => {
```
This approach is gone, so had to change this file as well.

mariusandra commented on frontend/src/toolbar/ToolbarApp.tsx:
```diff
@@ -1,42 +1,41 @@
-        ;(window as any)['__PHGTLB_ADD_STYLES__'] = (element: HTMLStyleElement) => addStyleElementToShadowRoot(element)
-    })
+    const didRender = useSecondRender(
+        props.disableExternalStyles
```
This `disableExternalStyles` prevents us from blocking and waiting for the stylesheet to open. It's only used inside storybook right now to bypass loading `toolbar.css` into the shadow root. The story lifts the styles into the shadow root in a different way.

mariusandra replied:
I tested that I could still see the toolbar's CSS locally with both webpack and esbuild, with both dev server and production builds.

mariusandra commented on package.json:
```diff
@@ -16,19 +16,27 @@
-        "start-http": "mkdir -p frontend/dist/ && cp -a frontend/public/* frontend/dist/ && npm run copy-scripts && webpack serve",
-        "start-https": "mkdir -p frontend/dist/ && cp -a frontend/public/* frontend/dist/ && npm run copy-scripts && webpack serve --https",
-        "start-docker": "mkdir -p frontend/dist/ && cp -a frontend/public/* frontend/dist/ && npm run copy-scripts && webpack serve --host 0.0.0.0",
+        "start": "concurrently -n ESBUILD,TYPEGEN -c yellow,green \"yarn start-http\" \"yarn run typegen:watch\"",
+        "start-http": "yarn start-http:esbuild",
+        "start-http:esbuild": "yarn clean && yarn copy-scripts && node frontend/build.mjs --dev",
+        "start-http:webpack": "mkdir -p frontend/dist/ && cp -a frontend/public/* frontend/dist/ && npm run copy-scripts && webpack serve",
+        "start-https": "yarn start-https:webpack",
+        "start-https:webpack": "yarn start-http:webpack --https",
+        "start-docker": "yarn start-docker:esbuild",
+        "start-docker:esbuild": "yarn start-http:esbuild --host 0.0.0.0",
+        "start-docker:webpack": "yarn start-http:webpack --host 0.0.0.0",
```
I tried not to be too disruptive here, and duplicated scripts wherever possible. With very quick changes in the scripts below we can switch the default build engine back to webpack.

mariusandra commented on webpack.config.js:
```diff
@@ -12,22 +12,6 @@ function createEntry(entry) {
     const commonLoadersForSassAndLess = [
         {
             loader: 'style-loader',
-            options:
-                entry === 'toolbar'
-                    ? {
-                          insert: function insertAtTop(element) {
-                              // tunnel behind the shadow root
-                              if (window.__PHGTLB_ADD_STYLES__) {
-                                  window.__PHGTLB_ADD_STYLES__(element)
-                              } else {
-                                  if (!window.__PHGTLB_STYLES__) {
-                                      window.__PHGTLB_STYLES__ = []
-                                  }
-                                  window.__PHGTLB_STYLES__.push(element)
-                              }
-                          },
-                      }
-                    : undefined,
```
Both webpack and esbuild now load the toolbar's styles via toolbar.css. This is no longer needed.

mariusandra reviewed (commented):
This is ready for a look! Who's up for the task? To aid the review, I added almost as many comments as there are changed files.

mariusandra committed:
Merge branch 'master' into esbuild

mariusandra commented:
I added a few reviewers. Definitely not expecting everyone to go through this line by line, but since it's a pretty big architecture change, would like everyone to be informed and voice concerns if there are any.

pauldambra commented on frontend/src/lib/components/DateFilter/DateFilterRange.tsx:
```diff
@@ -3,7 +3,7 @@ import dayjs from 'dayjs'
 
 import dayjsGenerateConfig from 'rc-picker/lib/generate/dayjs'
-import generatePicker from 'antd/es/date-picker/generatePicker'
+import generatePicker from 'antd/lib/date-picker/generatePicker'
```
The antd documentation doesn't use one consistently either. [Their package.json](https://github.com/ant-design/ant-design/blob/master/package.json#L45) points to `antd/es/` in the module field which _supposedly_ helps bundlers (see https://github.com/rollup/rollup/wiki/pkg.module and https://webpack.js.org/guides/author-libraries/#final-steps)

It would be interesting to test if this has any impact on bundle speed or size... 

But I don't think we should block this PR on that - ideally, our tools should be clever enough to cope with us importing from any of the antd build folders so that we fall into the pit of success

pauldambra commented on frontend/src/vars.scss:
```diff
@@ -129,7 +129,7 @@ $top_nav_height: 50px;
 }
 
-// Z-indexes
+// Z-indexes, sync with vars.ts
```
The need to make the same change in two places feels like a potential source of hard to diagnose bugs. I don't know what the solution is, but feels like this needs follow on work to resolve

pauldambra commented on frontend/utils.mjs:
```diff
@@ -0,0 +1,279 @@
+                    console.log('⌛️ Waiting for build to complete...')
+                    ifPaused.logged = true
+                    await ifPaused
+                    await new Promise((r) => setTimeout(r, 400))
```
I wonder if some of these GitHub comments would live better in the file itself

hazzadous replied:
💯  if they're useful to clarify the PR, they'll probably be useful for anyone reading this code in a dev context as well.

pauldambra commented on frontend/utils.mjs:
```diff
@@ -0,0 +1,279 @@
+                }
+                next()
+            },
+            createProxyMiddleware((pathname) => !!backendUrls.find((u) => pathname.startsWith(u)), {
+                target: backend,
+                changeOrigin: true,
+                logLevel: 'warn',
+            }),
```
I wonder if some of these GitHub comments would live better in the file itself

pauldambra commented on frontend/utils.mjs:
```diff
@@ -0,0 +1,279 @@
+                            'Content-Type': 'text/html',
+                        })
+                        res.end(
+                            `Can not access <a href="${backend}">${backend}</a>. Is the PostHog Django app running?`
```
love this kind of helpful error message 🙌

pauldambra commented:
Live reload isn't working for react component changes for me...

Steps taken

* checked out branch
* `nvm use v14`
* `yarn`
* restart Django server
* run `yarn start`
* load the site
* change a react component
* see that the page doesn't reload

mariusandra committed:
move live server injection into django

mariusandra committed:
fix undefined bug

mariusandra committed:
simplify setup to work with injection directly in http://localhost:8000 (no proxying needed on :8234)

mariusandra committed:
add comments

mariusandra commented:
Thanks for the review @pauldambra ! I added a few comments to `utils.mjs` and also greatly simplified the setup. The way this PR was set up now, to benefit from live reloads you had to use http://localhost:8234 for development. That would then proxy the result to django (port 8000) to get the `index.html` that would contain the app context payload.. and then inject some HTML to the proxied code.

If I just inject the reload code inside django, I can remove 1) the proxying of random urls to index.html, 2) the list of backend urls to not proxy, 3) the two injection middlewares, 4) the proxying middleware npm package 5) any knowledge the frontend build system has about the backend, 6) probably even some CO2 from the atmosphere.

So that's what I did. And as a side-effect, nobody has to change their workflow to run http://localhost:8234/ (and we avoided the issue to choose a new port!)

hazzadous commented:
> I had to build a new dev server since esbuild doesn't come with any

I haven't really looked into how applicable it is, but I thought it did: https://how-to.dev/how-to-set-up-a-dev-server-with-esbuild

mariusandra commented:
> > I had to build a new dev server since esbuild doesn't come with any
> I haven't really looked into how applicable it is, but I thought it did: https://how-to.dev/how-to-set-up-a-dev-server-with-esbuild

True, there is something, and I tried a bunch of approaches, but since we're building 3 apps at once, we'd need to run 3 servers on 3 ports, split the env `JS_URL` in three, etc. There was nothing that fit the bill. Plus, I really want to save the extra 2 seconds by starting reloads as soon as the build starts, not when it completes. Just like with webpack. Nothing provided this either.

Twixes committed:
Fix `fse` usage

I was getting this otherwise:

$ node frontend/build.mjs
file:///Users/twixes/Developer/posthog/frontend/utils.mjs:46
    fse.copySync(srcDir, destDir, { overwrite: true }, function (err) {
        ^

TypeError: fse.copySync is not a function
    at copyPublicFolder (file:///Users/twixes/Developer/posthog/frontend/utils.mjs:46:9)
    at file:///Users/twixes/Developer/posthog/frontend/build.mjs:5:1
    at ModuleJob.run (internal/modules/esm/module_job.js:146:23)
    at async Loader.import (internal/modules/esm/loader.js:165:24)
    at async Object.loadESM (internal/process/esm_loader.js:68:5)

Twixes committed:
Mock `process` for VFile used by ReactMarkdown

I was getting this otherwise:

core.js:55 Uncaught ReferenceError: process is not defined
    at new VFile (core.js:55)
    at VFile (core.js:49)
    at Function.parse (index.js:273)
    at ReactMarkdown2 (react-markdown.js:42)
    at renderWithHooks (react-dom.development.js:14803)
    at mountIndeterminateComponent (react-dom.development.js:17482)
    at beginWork (react-dom.development.js:18596)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:188)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237)
    at invokeGuardedCallback (react-dom.development.js:292)

Twixes commented on frontend/src/vars.ts:
```diff
@@ -0,0 +1,2 @@
+// z-indexes from vars.scss
```
So we can probably remove `:export { ... }` from `vars.scss`?

mariusandra replied:
🦔 💨

Twixes committed:
Mock `process.env` for VFile used by ReactMarkdown

I was getting this otherwise:

platform.ts:73 Uncaught TypeError: Cannot read properties of undefined (reading 'ENABLE_VSCODE_BROWSER_CODE_LOADING')
    at platform.ts:73
    at platform.ts:79
    at Function.r._invokeFactory (loader.js:1118)
    at r.complete (loader.js:1128)
    at r._onModuleComplete (loader.js:1754)
    at r._resolve (loader.js:1714)
    at r.defineModule (loader.js:1357)
    at _ (loader.js:1804)
    at numbers.ts:10
    at fake:1

Twixes reviewed (commented):
Had to make two fixes (e2c1bd2d22b60924ba22c66d8c0a2a57e7664f47, 9eca47d38fa0a90a06f591aa5ee3046f6c6c599d, cb19ae3f5c12360623100a935d57e80b3ae8c7ec) but this works fantastically well for me now. Just left a couple of clarification questions, beyond that I'd love to get this merged.

mariusandra committed:
pass the heavy appScenes to sceneLogic through props via App.tsx

mariusandra committed:
remove sceneProxyLogic

mariusandra commented on frontend/src/scenes/sceneLogic.ts:
```diff
@@ -15,7 +15,11 @@ import { SceneExport, Params, Scene, SceneConfig, SceneParams, LoadedScene } fro
 
+// Don't import and use this logic directly as via "scenes/scenes" it pulls in all scenes with
+// their dynamic imports, and hence confuses bundlers.
+// Instead use sceneProxyLogic.
```
Nothing will crash, and no bundle will grow out of control. What happened for me was `eventUsageLogic` was linked from every one of our 3 bundles (app, toolbar, shared), and since it linked to sceneLogic, all the 3 bundles reloaded on any change to any file. If that happens again, it'll be quick to spot and fix.

I reworked this now, and pass the `scenes`  to `sceneLogic` via `props`, not directly through an `import`. This ties all the scenes to `App.tsx` instead, which isn't really imported from anywhere.

mariusandra committed:
remove exported variables

mariusandra commented:
@Twixes Thanks for the fixes! I improved some of the things you suggested, but ain't nobody gonna touch the squeak!

mariusandra committed:
fix sceneLogic test

mariusandra committed:
Merge branch 'master' into esbuild

mariusandra committed:
Merge branch 'master' into esbuild
